enterprise/backend/src/metabase_enterprise/semantic_search/gate.clj

```diff
@@ -0,0 +1,90 @@
+  ;; NOTE: we depend on the UPDATE behaviour described here: https://www.postgresql.org/docs/current/transaction-iso.html#XACT-READ-COMMITTED
+  ;; i.e (the WHERE must be re-evaluated against the UPDATED row if another transaction commits concurrently).
+  (jdbc/with-transaction [tx pgvector {:isolation :read-committed}]
+    (when (seq gate-documents)
+      (let [{:keys [gate-table-name]} index-metadata
+            upsert-q   {:insert-into   [(keyword gate-table-name)]
+                        ;; sort to ensure locks are acquired predictably
+                        :values        (sort-by :id gate-documents)
+                        :on-conflict   [:id]
+                        :do-update-set {:updated_at :excluded.updated_at
+                                        :document   :excluded.document
+                                        :gated_at   [:clock_timestamp]}
+                        :where         [:< :updated_at :excluded.updated_at]}
+            upsert-sql (sql/format upsert-q :quoted true)]
+        (jdbc/execute! tx ["SET LOCAL statement_timeout = ?" (format "%ss" gate-write-timeout-seconds)])
+        (jdbc/execute! tx upsert-sql)
+        nil))))
+
+(defn poll
+  "Find document (gate table) ids that are highly likely to have been updated. Due to commit lag, if called with a watermark
+  close to the current postgres clock value, you will see some records with timestamps <= your watermark.
+  Therefore, any downstream processing must be idempotent - duplicate delivery is expected and intended.
+
+  Expect to receive a map with:
+  - `:poll-time` the postgres clock value as-of this poll
+  - `:update-candidates` a vector of [{:id, :gated_at}] maps.
+
+  Options:
+  - :limit (1000)
+  - :lag-tolerance-seconds (should be at least gate-write-timeout-seconds + some buffer for timeout variation)
+
+  Call (next-watermark) on the result to determine where to poll next."
+  [pgvector index-metadata watermark & {:keys [lag-tolerance-seconds
+                                               limit]
+                                        :or {lag-tolerance-seconds (* 2 gate-write-timeout-seconds) ; heuristic: still depends on postgres enforcing timeouts. We might still need a slow-pass over everything occasionally in the future.
+                                             limit 1000}}]
+  (let [poll-q            {:union-all
+                           [{:select [[-1 :id]
+                                      [[:clock_timestamp] :gated_at]]} ; return pgs clock value, to use as the :poll-time
+                            {:select [:id, :gated_at]
+                             :from   [(keyword (:gate-table-name index-metadata))]
+                             ;; the earliest timestamp where we might expect to find new documents
+                             :where  [:gated_at :>= [:min
+                                                     ;; lag our last-poll slightly, as we may not have seen everything
```

Comment:
After sleep this makes less sense, cant we just have the last poll - lag given its one clock (drop the `min`).

Only thing the :last-seen expr does in this `min` is make it so lag configuration is irrelevant when it is < the confidence window.

Makes stuff more confusing IMO with little beneft